datacite-delete-draft-doi-handler/src/main/java/no/unit/nva/datacite/handlers/resource/ExternalUpdatesEventHandler.java

```diff
@@ -0,0 +1,136 @@
+            S3Driver.defaultS3Client().build(),
+            defaultDoiClient());
+    }
+
+    protected ExternalUpdatesEventHandler(
+        Environment environment, S3Client s3Client, DoiClient doiClient) {
+        this.s3Driver = new S3Driver(s3Client, environment.readEnv(EVENTS_BUCKET_NAME_ENV));
+        this.doiClient = doiClient;
+    }
+
+    @Override
+    public Void handleRequest(SQSEvent sqsEvent, Context context) {
+        Optional.ofNullable(sqsEvent.getRecords()).stream()
+            .flatMap(List::stream)
+            .map(ExternalUpdatesEventHandler::parseEventReference)
+            .filter(Objects::nonNull)
+            .forEach(this::processPayload);
+        return null;
+    }
+
+    private void processPayload(EventReference eventReference) {
+        verifyTopic(eventReference);
+        var updateEvent = getEventBodyFromS3(eventReference);
+        verifyAction(updateEvent);
+        deleteDraftDoiIfPresent(updateEvent);
+    }
+
+    private void deleteDraftDoiIfPresent(UpdateEvent updateEvent) {
+        var customerId = updateEvent.oldData().publisher().id();
+        if (nonNull(updateEvent.oldData().doi())) {
+            var doi = Doi.fromUri(updateEvent.oldData().doi());
+            try {
+                var doiState = doiClient.getDoi(customerId, doi);
+                if (State.DRAFT.equals(doiState.getState())) {
+                    doiClient.deleteDraftDoi(customerId, doi);
+                    logger.info("Deleted draft DOI {} as resource {} was deleted.", doi,
+                                updateEvent.oldData().identifier());
+                }
+            } catch (ClientException e) {
+                throw new EventHandlingException("Failed to look up doi", e);
+            }
+        }
+    }
+
+    private static void verifyAction(UpdateEvent updateEvent) {
+        if (!REMOVE_ACTION.equals(updateEvent.action())) {
+            logger.error("Received unknown action in s3 event: {}", updateEvent.action());
+            throw new EventHandlingException("Unknown action in s3 event data. Expected REMOVE!");
+        }
+    }
+
+    private UpdateEvent getEventBodyFromS3(EventReference eventReference) {
+        var event = s3Driver.readEvent(eventReference.getUri());
+        return attempt(() -> JsonUtils.dtoObjectMapper.readValue(event, UpdateEvent.class))
+                   .orElseThrow(this::logAndThrow);
+    }
+
+    private static void verifyTopic(EventReference eventReference) {
+        if (!RESOURCE_DELETED_TOPIC.equals(eventReference.getTopic())) {
+            logger.error(
+                "Got external update event with message on unknown topic {}", eventReference.getTopic());
+            throw new EventHandlingException(
+                "Received external update event with message on unknown topic!");
+        }
+    }
+
+    private RuntimeException logAndThrow(Failure<UpdateEvent> updateEventFailure) {
+        final Throwable cause = updateEventFailure.getException();
```

Comment:
var?

Reply:
This actually made codacy understand that updateEventFailure.getException() is a throwable and not just a var-arg object for string formatting with missing `{}` .